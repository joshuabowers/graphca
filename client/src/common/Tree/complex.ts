import { Complex } from './Expression';
// import { Complex } from './Constant/Complex'

export function complex(a: number, b: number) {
  return new Complex(a, b)
}
