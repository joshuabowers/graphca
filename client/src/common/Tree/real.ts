import { Real } from "./Expression";
// import { Real } from './Constant/Real'

export function real(value: number) {
  return new Real(value)
}
