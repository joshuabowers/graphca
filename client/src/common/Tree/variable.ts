import { Base } from './Expression'

export class Variable extends Base {
  readonly $kind = 'Variable'
  constructor(readonly name: string, readonly value?: Base) { super() }
}

export function variable(name: string, value?: Base) {
  return new Variable(name, value)
}
