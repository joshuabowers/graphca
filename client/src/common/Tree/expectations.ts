import { method, multi, Multi, _ } from '@arrows/multimethod'
import { is } from './is'
import { Real } from './real'
import { Complex } from './complex'

export type ExpectCloseTo = Multi 
  & ((actual: Real, expected: Real, precision: number) => void)
  & ((actual: Complex, expected: Complex, precision: number) => void)

export const expectCloseTo: ExpectCloseTo = multi(
  method(
    [is(Real), is(Real), _], 
    (actual: Real, expected: Real, precision: number) => 
      expect(actual.value).toBeCloseTo(expected.value, precision)
  ),
  method(
    [is(Complex), is(Complex), _],
    (actual: Complex, expected: Complex, precision: number) => {
      expect(actual.a).toBeCloseTo(expected.a, precision)
      expect(actual.b).toBeCloseTo(expected.b, precision)
    }
  )
)
