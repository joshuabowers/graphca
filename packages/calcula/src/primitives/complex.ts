import { Species } from "../utility/tree"
import { Complex, primitive } from "../closures/primitive"
import { isNumberTuple } from "../utility/valuePredicates"
export { Complex }

export const [complex, isComplex, $complex] = 
  primitive<[number, number], {a: number, b: number}, Complex>(
  isNumberTuple,
  ([a, b]) => ({a, b}),
  Species.complex
)(
  r => [r.value, 0],
  c => [c.a, c.b],
  b => [b.value ? 1 : 0, 0]
)()

export const ComplexInfinity = complex([Infinity, NaN])
