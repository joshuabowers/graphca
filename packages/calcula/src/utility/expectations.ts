import { method, multi, Multi, _ } from '@arrows/multimethod'
import { isWriter, Writer } from '../monads/writer'
import { TreeNode } from './tree'
import { Real, Complex, isReal, isComplex } from '../primitives'

export type ExpectCloseTo = Multi 
  & ((actual: Writer<Real>, expected: Writer<Real>, precision: number) => void)
  & ((actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => void)

export const expectCloseTo: ExpectCloseTo = multi(
  method(
    [isReal, isReal, _], 
    (actual: Writer<Real>, expected: Writer<Real>, precision: number) => 
      expect(actual.value.value).toBeCloseTo(expected.value.value, precision)
  ),
  method(
    [isComplex, isComplex, _],
    (actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => {
      expect(actual.value.a).toBeCloseTo(expected.value.a, precision)
      expect(actual.value.b).toBeCloseTo(expected.value.b, precision)
    }
  )
)

type Input = TreeNode|Writer<TreeNode>|[Input, Input]
export type Operation = [Input, string]

export const expectWriter = <
  Actual extends TreeNode, 
  Expected extends TreeNode|Writer<TreeNode>
>(
  actual: Writer<Actual>
) => (expected: Expected, ...operations: Operation[]) => {
  expect(actual).toEqual({
    value: isWriter(expected) ? expected.value : expected,
    log: operations.map(([input, action]) => ({
      input: isWriter(input) 
        ? input.value 
        : (Array.isArray(input) 
          ? input.map(i => isWriter(i) ? i.value : i)
          : input), 
      action
    }))
  })
}
