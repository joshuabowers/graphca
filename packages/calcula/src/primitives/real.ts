import { Species } from '../utility/tree'
import { Real, primitive } from '../closures/primitive'
import { isNumericOrString } from '../utility/valuePredicates'
import { numeric } from '../utility/numeric'
import { nan } from './nan'
export { Real }

export const [real, isReal, $real] = 
  primitive<[number|string], number, Real>(
  isNumericOrString,
  value => Number(value),
  Species.real,
  r => numeric(r.raw)
)(
  r => [r.raw],
  c => [c.raw.a],
  b => [b.raw ? 1 : 0]
)(
  when => [
    when(v => Number.isNaN(v), [nan, 'incalculable'])
  ]
)

export const EulerMascheroni = real(0.57721566490153286060)
