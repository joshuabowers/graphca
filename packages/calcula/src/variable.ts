import { Writer, unit } from './monads/writer'
import { TreeNode, Clades, Species, isSpecies } from './utility/tree'
import { isNil, nil } from './primitives/nil'
import { rule, process, resolve } from './utility/rule'

export type Variable = TreeNode & {
  readonly clade: Clades.variadic,
  readonly species: Species.variable,
  readonly name: string,
  readonly value: Writer<TreeNode>
}

export const isVariable = isSpecies<Variable>(Species.variable)

export const variable = (
  name: string, value: Writer<TreeNode> = nil
): Writer<Variable> => 
  ({
    value: {
      clade: Clades.variadic,
      species: Species.variable,
      name, value
    },
    log: [{
      input: process`${name}`, 
      rewrite: isNil(value) ? resolve`${name}` : process`${value}`,
      action: isNil(value) 
        ? 'referenced variable' : 'referenced variable with value'
    }] 
  })

export type Scope = Map<string, Writer<Variable>>

export const isScope = (v: unknown): v is Scope =>
  typeof v === 'object' && v !== null
  && ['get', 'set', 'has', 'delete'].every(m => m in v)

type Entries = Iterable<readonly [string, Writer<Variable>]>

export const scope = (entries: Entries = []): Scope => 
  new Map<string, Writer<Variable>>(entries)

export const assign = (name: string, value: Writer<TreeNode>, scope?: Scope) => {
  const v = variable(name, value)
  if(scope){
    if(isNil(value)){ scope.delete(name) }
    else { scope.set(name, v) }
  }
  return v
}
