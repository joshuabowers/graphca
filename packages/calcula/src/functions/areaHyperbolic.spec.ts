import { expectCloseTo } from './expectations'
import { real } from './real'
import { complex } from './complex'
import { variable } from './variable'
import { 
  AreaHyperbolicCosine, AreaHyperbolicSine, AreaHyperbolicTangent,
  AreaHyperbolicSecant, AreaHyperbolicCosecant, AreaHyperbolicCotangent,
  acosh, asinh, atanh, asech, acsch, acoth 
} from './areaHyperbolic'

describe('acosh', () => {
  it('calculates the area hyperbolic cosine of a real value', () => {
    expectCloseTo(acosh(real(1)), real(0), 10)
  })

  it('calculates the area hyperbolic cosine of a complex number', () => {
    expectCloseTo(acosh(complex(2, 1)), complex(1.469351744368, 0.507356303217), 10)
  })

  it('generates an AreaHyperbolicCosine node of a variable expression', () => {
    expect(acosh(variable('x'))).toEqual(new AreaHyperbolicCosine(variable('x')))
  })
})

describe('asinh', () => {
  it('calculates the area hyperbolic sine of a real value', () => {
    expectCloseTo(asinh(real(1)), real(0.88137358701), 10)
  })

  it('calculates the area hyperbolic sine of a complex number', () => {
    expectCloseTo(asinh(complex(0, 1)), complex(0, 1.57079632679), 7)
  })

  it('generates an AreaHyperbolicSine node of a variable expression', () => {
    expect(asinh(variable('x'))).toEqual(new AreaHyperbolicSine(variable('x')))
  })
})

describe('atanh', () => {
  it('calculates the area hyperbolic tangent of a real value', () => {
    expectCloseTo(atanh(real(1)), real(Infinity), 10)
  })

  it('calculates the area hyperbolic tangent of a complex number', () => {
    expectCloseTo(atanh(complex(0, 1)), complex(0, 0.78539816339), 10)
  })

  it('generates an AreaHyperbolicTangent node of a variable expression', () => {
    expect(atanh(variable('x'))).toEqual(new AreaHyperbolicTangent(variable('x')))
  })
})

describe('asech', () => {
  it('calculates the area hyperbolic secant of a real value', () => {
    expectCloseTo(asech(real(1)), real(0), 10)
  })

  it('calculates the area hyperbolic secant of a complex number', () => {
    expectCloseTo(asech(complex(0, 1)), complex(0.88137358701, -1.57079632679), 10)
  })

  it('generates an AreaHyperbolicSecant node of a variable expression', () => {
    expect(asech(variable('x'))).toEqual(new AreaHyperbolicSecant(variable('x')))
  })
})

describe('acsch', () => {
  it('calculates the area hyperbolic cosecant of a real value', () => {
    expectCloseTo(acsch(real(1)), real(0.88137358701), 10)
  })

  it('calculates the area hyperbolic cosecant of a complex number', () => {
    expectCloseTo(acsch(complex(0, 1)), complex(0, -1.57079632679), 7)
  })

  it('generates an AreaHyperbolicCosecant node of a variable expression', () => {
    expect(acsch(variable('x'))).toEqual(new AreaHyperbolicCosecant(variable('x')))
  })
})

describe('acoth', () => {
  it('calculates the area hyperbolic cotangent of a real value', () => {
    expectCloseTo(acoth(real(1)), real(Infinity), 10)
  })

  it('calculates the area hyperbolic cotangent of a complex number', () => {
    expectCloseTo(acoth(complex(0, 1)), complex(0, -0.78539816339), 10)
  })

  it('generates an AreaHyperbolicCotangent node of a variable expression', () => {
    expect(acoth(variable('x'))).toEqual(new AreaHyperbolicCotangent(variable('x')))
  })
})
