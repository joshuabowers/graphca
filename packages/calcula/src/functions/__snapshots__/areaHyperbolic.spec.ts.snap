// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`acosh generates an AreaHyperbolicCosine node of a variable expression 1`] = `
[
  [
    "acosh(x)",
    "identified area hyperbolic cosine",
  ],
  [
    "acosh([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "acosh({x})",
    "processed argument",
  ],
  [
    "acosh(x)",
    "created area hyperbolic cosine",
  ],
]
`;

exports[`acoth generates an AreaHyperbolicCotangent node of a variable expression 1`] = `
[
  [
    "acoth(x)",
    "identified area hyperbolic cotangent",
  ],
  [
    "acoth([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "acoth({x})",
    "processed argument",
  ],
  [
    "acoth(x)",
    "created area hyperbolic cotangent",
  ],
]
`;

exports[`acsch generates an AreaHyperbolicCosecant node of a variable expression 1`] = `
[
  [
    "acsch(x)",
    "identified area hyperbolic cosecant",
  ],
  [
    "acsch([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "acsch({x})",
    "processed argument",
  ],
  [
    "acsch(x)",
    "created area hyperbolic cosecant",
  ],
]
`;

exports[`asech calculates the area hyperbolic secant of a real value 1`] = `
[
  [
    "asech(2)",
    "identified area hyperbolic secant",
  ],
  [
    "asech([2])",
    "processing argument",
  ],
  [
    "2",
    "created real",
  ],
  [
    "asech({2})",
    "processed argument",
  ],
  [
    "asech(2)",
    "real area hyperbolic secant",
  ],
  [
    "acosh((2^-1))",
    "identified area hyperbolic cosine",
  ],
  [
    "acosh([(2^-1)])",
    "processing argument",
  ],
  [
    "(2^-1)",
    "identified exponentiation",
  ],
  [
    "([2]^-1)",
    "processing left operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "({2}^[-1])",
    "processed left operand; processing right operand",
  ],
  [
    "-1",
    "created real",
  ],
  [
    "(2^{-1})",
    "processed right operand",
  ],
  [
    "(2^-1)",
    "real exponentiation",
  ],
  [
    "0.5",
    "created real",
  ],
  [
    "acosh({0.5})",
    "processed argument",
  ],
  [
    "acosh(0.5)",
    "real area hyperbolic cosine",
  ],
  [
    "NaN",
    "created real",
  ],
]
`;

exports[`asech generates an AreaHyperbolicSecant node of a variable expression 1`] = `
[
  [
    "asech(x)",
    "identified area hyperbolic secant",
  ],
  [
    "asech([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "asech({x})",
    "processed argument",
  ],
  [
    "asech(x)",
    "created area hyperbolic secant",
  ],
]
`;

exports[`asinh generates an AreaHyperbolicSine node of a variable expression 1`] = `
[
  [
    "asinh(x)",
    "identified area hyperbolic sine",
  ],
  [
    "asinh([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "asinh({x})",
    "processed argument",
  ],
  [
    "asinh(x)",
    "created area hyperbolic sine",
  ],
]
`;

exports[`atanh generates an AreaHyperbolicTangent node of a variable expression 1`] = `
[
  [
    "atanh(x)",
    "identified area hyperbolic tangent",
  ],
  [
    "atanh([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "atanh({x})",
    "processed argument",
  ],
  [
    "atanh(x)",
    "created area hyperbolic tangent",
  ],
]
`;
